src/mqtt/protocol/mqtt/mqtt_client.c

```diff
@@ -0,0 +1,1110 @@
+	nni_aio_close(&p->send_aio);
+	nni_aio_close(&p->recv_aio);
+	mqtt_sock_close_work_queue(s, &s->send_queue);
+	mqtt_sock_close_work_queue(s, &s->recv_queue);
+	nni_id_map_foreach(&p->send_unack, mqtt_close_unack_work_cb);
+	nni_id_map_foreach(&p->recv_unack, mqtt_close_unack_work_cb);
+	nni_mtx_unlock(&s->mtx);
+
+	nni_atomic_set_bool(&p->closed, true);
+}
+
+static uint16_t
+mqtt_pipe_get_next_packet_id(mqtt_pipe_t *p)
+{
+	int packet_id;
+	do {
+		packet_id = nni_atomic_get(&p->next_packet_id);
+	} while (
+	    !nni_atomic_cas(&p->next_packet_id, packet_id, packet_id + 1));
+	return packet_id & 0xFFFF;
+}
+
+static inline void
+mqtt_pipe_recv_msgq_putq(mqtt_pipe_t *p, nni_msg *msg)
+{
+	if (0 != nni_lmq_put(&p->recv_messages, msg)) {
+		// resize to ensure we do not lost messages
+		// add option to drop messages
+		if (0 !=
+		    nni_lmq_resize(&p->recv_messages,
+		        nni_lmq_len(&p->recv_messages) * 2)) {
+			// drop the message when no memory available
+			nni_msg_free(msg);
+			return;
+		}
+		nni_lmq_put(&p->recv_messages, msg);
+	}
+}
+
+// Keep alive timer callback to send ping request.
+static void
+mqtt_keep_alive_cb(void *arg)
+{
+	work_t *     work = arg;
+	mqtt_sock_t *s    = work->mqtt_sock;
+
+	nni_mtx_lock(&s->mtx);
+
+	mqtt_pipe_t *p = s->mqtt_pipe;
+
+	if (NULL == p || nni_atomic_get_bool(&p->closed)) {
+		return;
+	}
+
+	// keep alive time out
+	// if there is work in the send queue, no need to send PINGREQ
+	if (nni_list_empty(&s->send_queue)) {
+		// no work in send queue, then send a PINGREQ
+		nni_list_append(&s->send_queue, &p->ping_work);
+		mqtt_send_start(s);
+	}
+
+	nni_mtx_unlock(&s->mtx);
+	work_timer_schedule(&p->ping_work);
+}
+
+// Timer callback, we use it for retransmitting.
+static void
+mqtt_timer_cb(void *arg)
+{
+	work_t *     work = arg;
+	mqtt_sock_t *s    = work->mqtt_sock;
+
+	nni_mtx_lock(&s->mtx);
+
+	mqtt_pipe_t *p = s->mqtt_pipe;
+
+	if (NULL == p || nni_atomic_get_bool(&p->closed)) {
+		return;
+	}
+
+	if (work->ntrial < nni_atomic_get(&s->ttl)) {
+		// try again
+		nni_list_append(&s->send_queue, work);
+		if (nni_list_first(&s->send_queue) == work) {
+			mqtt_send_start(s);
+		}
+	} else {
+		// reach max retransmitting, quit
+		nni_aio_finish_error(work->user_aio, NNG_EAGAIN);
+		nni_id_remove(&p->send_unack, work->packet_id);
+		mqtt_sock_free_work(s, work);
+	}
+
+	nni_mtx_unlock(&s->mtx);
+}
+
+static void
+mqtt_send_cb(void *arg)
+{
+	mqtt_pipe_t *p = arg;
+	mqtt_sock_t *s = p->mqtt_sock;
+	work_t *     work;
+	bool         scheduler = false;
+
+	nni_mtx_lock(&s->mtx);
+
+	work = nni_list_first(&s->send_queue); // will not be NULL
+	nni_list_remove(&s->send_queue, work);
+
+	if (nni_aio_result(&p->send_aio) != 0) {
+		// We failed to send... clean up and deal with it.
+		nni_msg_free(nni_aio_get_msg(&p->send_aio));
+		nni_aio_set_msg(&p->send_aio, NULL);
+		nni_mtx_unlock(&s->mtx);
+		nni_pipe_close(p->pipe);
+		return;
+	}
+
+	if (nni_atomic_get_bool(&s->closed) ||
```

Comment:
Suspect these changes should not be needed -- probably the callback would be canceled with NNG_ECLOSED in that case.

Reply:
I am not sure either. lack a way to identify it.